### PR TITLE
Async listener and private channel events

### LINKED ISSUE #1294: Inconsistent Use of Promises
`Channel` contains this, promised API call:

```
public addContextListener(contextType: string | null, handler: ContextHandler): Promise<Listener>;
```

`PrivateChannel` however, has unpromised API calls, which also return listeners:

```
interface  PrivateChannel extends Channel {
  // methods
  onAddContextListener(handler: (contextType?: string) => void): Listener;
  onUnsubscribe(handler: (contextType?: string) => void): Listener;
  onDisconnect(handler: () => void): Listener;
  disconnect(): void;
}
```

... which lead to race conditions in my tests, sadly.  I doubt we'll ever fix this but raising nonetheless.

### LINKED ISSUE #1300: Make Listener.unsubscribe() async for consistency
### Question Area

 - [ ] App Directory
 - [x] API
 - [ ] Context Data
 - [ ] Intents
 - [ ] Use Cases
 - [ ] Other

### Question

Nearly all of the functions in the api are async and return a `Promise`. `Listener.unsubscribe()` is not.

I can see that it could well be a fire and forget function but within `BrowserTypes` we have `ContextListenerUnsubscribeRequest`, `ContextListenerUnsubscribeResponse`, `IntentListenerUnsubscribeRequest` and `IntentListenerUnsubscribeResponse`.

This leads me to assume that when we call `unsubscribe` we need to dispatch a Request message and wait for a Response message. In the meantime we probably aren't unsubscribed.

Is this assumption incorrect? Should we unsubscribe immediately before we get the response message? In which case what is the point of the response message?
How do we handle errors with the unsubscribe process?

### PR DESCRIPTION
**Pull request by kriswest**
resolves #1294 
resolves #1300 

* Makes `Listener.unsubscribe()` async (the return type is changed from `void` to `Promise<void>`) for consistency with the rest of the API.
* Adds an `async` `addEventListener` function to the `PrivateChannel` API to replace the deprecated, synchronous `onAddContextListener`, `onUnsubscribe` and `onDisconnect` functions and to keep consistency with the DesktopAgent API.
* Deprecates `PrivateChannel`'s synchronous `onAddContextListener`, `onUnsubscribe` and `onDisconnect` functions in favour of an `async` `addEventListener` function consistent with the one added to `DesktopAgent` in #1207
* Adds missing Changle log entry for #1207

See previews at:
- https://deploy-preview-1305--fdc3.netlify.app/docs/next/api/ref/Types#listener
- https://deploy-preview-1305--fdc3.netlify.app/docs/next/api/ref/Events
- https://deploy-preview-1305--fdc3.netlify.app/docs/next/api/ref/PrivateChannel

### CODE REVIEW

**Comment by kriswest** [comment]
@Roaders @robmoffat This PR addresses the async API call issues you both raised. Let me know what you think.

**Comment by robmoffat** [comment]
I like what you did here to unify the event listener stuff.  I guess we're committing to this for 2.2 then? 
We haven't actually discussed the listener changes in any meetings yet though

**Comment by kriswest** [comment]
no we haven't. that would need to happen at the next meeting, having the pr and docs preview in hand makes it a lot easier to show the change however! 

k

**Comment by Roaders** [comment]
This looks good to me. I don't think that there is any easy way to test this in my local code other than building the `.tgz` myself and manually installing it in my project right?

**Comment by kriswest** [comment]
Correct. Its easy enough todo however, checkout the relevant branch, run  `npm run pack` which will build the project and then pack the tarball for you. Then simply `npm install <path to tarball>` in the project where you want to use it.

**Comment by kriswest** [comment]
P.S. @Roaders  if you try to do that with the FDC3 for Web PR, note that I haven't got around to adding BrowserTypes.to the exports in index.ts yet. It'll probably conflict with existing API exports so we'll probably need to export it as 'BrowserTypes' the same we do with BridgingTypes.

**Comment by kriswest** [comment]
@Roaders @bingenito @hughtroeger I'll put this on the SWG agenda to look at this week. Would love to get some reviews on it.

File: src/api/Events.ts
Comment (bingenito) [inline src/api/Events.ts:13]: Is there a good common base type or marker interface recommendation for this in other languages?

Comment (kriswest) [reply src/api/Events.ts:13]: Ah the dreaded union type again.

The event types are just strings really, we're just using the enums to govern the set as we do with errors - they are all that really differs. The type names are mutually exclusive between the fdc3 interface events and privateChannel interface events, but it felt like they should be different enums.

I haven't included a common anscestor for the two event types (`FDC3Event` & `PrivateChannelEvent`), which is I guess what you are after! I can go add that and have both extend it with specific enums for the types. That would mean EventHandler could use that ancestor here instead...

I'll go take a look at this - I suspect typescript enums are going to make the above more difficult than it should be... Might have to switch to string unions...

**Comment by kriswest** [comment]
Working on an update that remove the enums as then you can have a workable inheritance structure

File: src/api/Events.ts
Comment (kriswest) [reply src/api/Events.ts]: Checkout https://github.com/finos/FDC3/commit/c47b7b2c38c40444360870f6c684d00524982cd2 
Hopefully it'll alleviate your issue here by introducing a base ApiEvent type that the others restrict down to specific messages.

File: src/api/Events.ts
Comment (bingenito) [inline src/api/Events.ts:28]:
```suggestion
export type FDC3EventTypes = "userChannelChanged";
```

File: src/api/Events.ts
Comment (bingenito) [inline src/api/Events.ts:44]:
```suggestion
  readonly type: "userChannelChanged";
```

File: src/api/Events.ts
Comment (bingenito) [reply src/api/Events.ts:28]: As discussed, for all places, easiest to just mark the suggestion here

**Comment by kriswest** [comment]
Updated to keep event type names style consistent (camelCase) as agreed at SWG meeting 22/08/24. Also, consent was received to make these changes once they are passed through the review of maintainers/editors.

**Review by bingenito — DISMISSED** [review]

File: docs/api/ref/PrivateChannel.md
Comment (hughtroeger) [inline docs/api/ref/PrivateChannel.md:191]:
```suggestion
Deprecated in favour of the async `addEventListener("disconnect", handler)` function.
```

**Review by hughtroeger — APPROVED** [review]
One typo, otherwise looks good.

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [reply docs/api/ref/PrivateChannel.md]: Fixed, @hughtroeger or @bingenito pr @finos/fdc3-maintainers can you re-approve so I can merge pls? (There was a push so past reviews are invalidated)

**Review by bingenito — DISMISSED** [review]

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:33]: @bingenito, need to replicate this in the .NET block below

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:80]: @bingenito need to replicate switch to new function in .NET version below

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:91]: @bingenito need to replicate switch to new function in .NET version below

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:99]: @bingenito need to replicate switch to new function in .NET version below

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:162]: @bingenito need to replicate switch to new function in .NET version below

File: docs/api/ref/PrivateChannel.md
Comment (kriswest) [inline docs/api/ref/PrivateChannel.md:260]: @bingenito need to function definition here

**Review by kriswest — COMMENTED** [review]
This did not merge well with with the .NET PR, particularly in the PrivateChannel.md file - where some work is needed from @bingenito to fill in .NET examples. 

THere is also a syntax error somewhere in PrivateChannel.md that I can't find, that is preventing the site from compiling...

**Comment by bingenito** [comment]
> This did not merge well with with the .NET PR, particularly in the PrivateChannel.md file - where some work is needed from @bingenito to fill in .NET examples.
> 
> THere is also a syntax error somewhere in PrivateChannel.md that I can't find, that is preventing the site from compiling...

This was never implemented.  The pending work is under 1318 to done once this is merged.

**Comment by bingenito** [comment]
> This did not merge well with with the .NET PR, particularly in the PrivateChannel.md file - where some work is needed from @bingenito to fill in .NET examples.
> 
> THere is also a syntax error somewhere in PrivateChannel.md that I can't find, that is preventing the site from compiling...

It seems as though it is trying to transpile the markdown rather than build docs.

**Comment by kriswest** [comment]
> > This did not merge well with with the .NET PR, particularly in the PrivateChannel.md file - where some work is needed from @bingenito to fill in .NET examples.
> > THere is also a syntax error somewhere in PrivateChannel.md that I can't find, that is preventing the site from compiling...
> 
> It seems as though it is trying to transpile the markdown rather than build docs.

Looks like it - but I think theres a syntax error somewhere that is making it think that the code block is markup it needs to process. Its fine with other code blocks on that page so it's likely a stray or missing character I'm just not seeing...

**Comment by kriswest** [comment]
> > This did not merge well with with the .NET PR, particularly in the PrivateChannel.md file - where some work is needed from @bingenito to fill in .NET examples.
> > THere is also a syntax error somewhere in PrivateChannel.md that I can't find, that is preventing the site from compiling...
> 
> This was never implemented. The pending work is under #1318 to done once this is merged.

SHall I just go and add the tab but put Not implemented yet in it?

**Comment by bingenito** [comment]
Unless you think it is causing this issue I wouldn't bother as I'll do it with a PR after this merge.  I think there is a different issue here perhaps with mismatched tags that I'm looking at.  The syntax of that example is the same in the merged PR.

**Comment by bingenito** [comment]
That force push was for a git commit amend of my last commit message

**Review by bingenito — APPROVED** [review]

**Review by hughtroeger — APPROVED** [review]

**Comment by kriswest** [comment]
@hughtroeger no worries - thanks for getting to another one!

I spotted a syntax error (another stray space char):
image

Fixed, could you re-approve @hughtroeger ?

**Review by bingenito — APPROVED** [review]

**Pull request merged**
